File: metrics/cache_metrics.go
package metrics

import "reservoir/utils/atomics"

type cacheMetrics struct {
	CacheHits      atomics.Int64 `json:"cache_hits"`
	CacheMisses    atomics.Int64 `json:"cache_misses"`
	CacheErrors    atomics.Int64 `json:"cache_errors"`
	CacheEntries   atomics.Int64 `json:"cache_entries"`
	BytesCached    atomics.Int64 `json:"bytes_cached"`
	CleanupRuns    atomics.Int64 `json:"cleanup_runs"`
	BytesCleaned   atomics.Int64 `json:"bytes_cleaned"`
	CacheEvictions atomics.Int64 `json:"cache_evictions"`
}

func NewCacheMetrics() cacheMetrics {
	return cacheMetrics{
		CacheHits:      atomics.NewInt64(0),
		CacheMisses:    atomics.NewInt64(0),
		CacheErrors:    atomics.NewInt64(0),
		CacheEntries:   atomics.NewInt64(0),
		BytesCached:    atomics.NewInt64(0),
		CleanupRuns:    atomics.NewInt64(0),
		BytesCleaned:   atomics.NewInt64(0),
		CacheEvictions: atomics.NewInt64(0),
	}
}

File: webserver/dashboard/frontend/src/lib/api/objects/metrics/cache-metrics.ts
import { apiGet, type FetchFn } from "$lib/api/api-helpers";

export type CacheStorageMetrics = {
    type: "memory" | "file";
    entries: number;
    bytes: number;
    max_bytes: number;
    memory_cap_bytes?: number;
};

export type CacheMetrics = {
    cache_hits: number;
    cache_misses: number;
    cache_errors: number;
    cache_request_hits: number;
    cache_request_misses: number;
    cache_request_revalidations: number;
    cache_request_stales: number;
    cache_entries: number;
    bytes_cached: number;
    cleanup_runs: number;
    bytes_cleaned: number;
    cache_evictions: number;
    cache_hit_latency: number;
    cache_miss_latency: number;
    storage: CacheStorageMetrics;
};

export async function getCacheMetrics(fetchFn: FetchFn = fetch): Promise<Readonly<CacheMetrics>> {
    return apiGet<CacheMetrics>("/metrics/cache", fetchFn);
}
